torch/csrc/distributed/autograd/context/dist_autograd_context.cpp

```diff
@@ -13,16 +13,41 @@ int64_t DistAutogradContext::context_id() const {
 }
 
 void DistAutogradContext::addSendFunction(
-    const std::shared_ptr<SendRpcBackward>& func) {
+    const std::shared_ptr<SendRpcBackward>& func,
+    int64_t autograd_message_id) {
+  TORCH_INTERNAL_ASSERT(func != nullptr);
+
+  std::lock_guard<std::mutex> guard(lock_);
+  TORCH_INTERNAL_ASSERT(
+      sendAutogradFunctions_.find(autograd_message_id) ==
```

Comment:
What is the reason for autograd_functions to be keyed by autograd_message_id ?

Reply:
When a 'recv' function sends gradients to another node in the backward pass, we need to look up the appropriate 'send' function on that node (this is done using a combination of autograd_context_id and autograd_message_id).